File: src/main/java/com/ywh/problem/leetcode/easy/LeetCode344.java
package com.ywh.problem.leetcode.easy;

/**
 * 反转字符串
 * [字符串] [双指针]
 *
 * @author ywh
 * @since 2/16/2019
 */
public class LeetCode344 {

    /**
     * 注意区别于二分搜索的循环条件为 l <= h，反转问题的循环条件为 i < j
     *
     * Time: O(n), Space: O(n)
     *
     * @param s
     * @return
     */
    public String reverseString(String s) {
        if (s == null) {
            return null;
        }
        char[] c = s.toCharArray();
        for (int i = 0, j = c.length - 1; i < j; i++, j--) {
            char tmp = c[i];
            c[i] = c[j];
            c[j] = tmp;
        }
        return String.valueOf(c);
    }

}

File: python/src/problem/leetcode/easy/leetcode_344.py
# coding=utf-8
from typing import List


class Solution:
    """
    反转字符串
    """

    def reverse_string(self, s: List[str]) -> None:
        """
        Time: O(n), Space: O(n)
        :param s:
        :return:
        """
        left, right = 0, len(s) - 1
        while left < right:
            s[left], s[right] = s[right], s[left]
            left += 1
            right -= 1

